### ISSUE TITLE
logitGetter correct usage

### ISSUE THREAD

**Issue by annongitter** [KevinMusgrave/pytorch-metric-learning#255]
Greetings!

I have seen the docs for LogitGetter and I would like to use this in inference, but I am unsure if I have got the usage right. So, my structure looks something like this:

```
def test(model,logit_getter, test_loader):
   model.eval()
   with torch.no_grad():
        ## iterate through test_loader and do:
       embeddings = model(data)
       logits= logit_getter(embeddings) 
       probs=F.softmax(logits, dim=1)
       ## compare with target and log accuracy
   return None

model = SomeModel() # some model
lossfunc = losses..LargeMarginLoss(..) # some acceptable loss function with trainable params
...
for epoch in range(0, epochs):
    train(model=model,lossfunc=lossfunc)
    logit_getter =LogitGetter(classifier=lossfunc)
    # do tests with logitgetter, which does logits= logit_getter(embeddings) to compute output probabilities
    # like so: probs=F.softmax(logits, dim=1)
    test(model=model,logit_getter=logit_getter, test_loader=test_loader)

```

Is this a correct way to use `LogitGetter`? 

I have two other questions re LogitGetter:

- I see an issue here: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/221#issuecomment-715360459 and I was wondering if there is an issue when number of classes is equal to the embedding size?
- How are the logits exactly calculated? I can see that this should use the trainable parameters of the loss function, but I was unable to transverse through the code to locate where the logits are calculated for example for LargeMargin/ArcFace loss (I mean, it should be in a classifier layer I would have thought?). Any pointers to this would be great.

**Comment by KevinMusgrave** [comment]
> Is this a correct way to use `LogitGetter`?

Yes

> * I see an issue here: [#221 (comment)](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/221#issuecomment-715360459) and I was wondering if there is an issue when number of classes is equal to the embedding size?

If you don't pass in a ```transpose``` argument when you initialize LogitGetter, then during the forward pass, it decides whether or not to transpose the weight matrix based on the shape of the input embeddings and the shape of the weight matrix. If the weight matrix is square, then transposing and not transposing both work, so it can't make a correct decision. 

If you want to avoid this ambiguity, you can explicitly set the transpose flag, but it depends on the loss function:

```python
# for ArcFaceLoss, CosFaceLoss, SphereFaceLoss, LargeMarginSoftmaxLoss, NormalizedSoftmaxLoss
logit_getter = LogitGetter(classifier = loss_func, transpose = True)

# for ProxyAnchorLoss, ProxyNCALoss
logit_getter = LogitGetter(classifier = loss_func, transpose = False)
```

Obviously this isn't great because you have to know whether or not to transpose the weight matrix. I've created an issue to address this: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/256

> * How are the logits exactly calculated? I can see that this should use the trainable parameters of the loss function, but I was unable to transverse through the code to locate where the logits are calculated for example for LargeMargin/ArcFace loss (I mean, it should be in a classifier layer I would have thought?). Any pointers to this would be great.

LogitGetter contains a copy of the weight matrix of the loss function. When you pass in embeddings, it does ```distance(embeddings, weight_matrix)```, where ```distance``` is the distance metric used in the loss function. For example, ```ArcFaceLoss``` uses cosine similarity, and NormalizedSoftmaxLoss uses the dot product.

**Comment by annongitter** [comment]
> > Is this a correct way to use `LogitGetter`?
> 
> Yes

Thank you for clarifying this.

> 
> > * I see an issue here: [#221 (comment)](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/221#issuecomment-715360459) and I was wondering if there is an issue when number of classes is equal to the embedding size?
> 
> If you don't pass in a `transpose` argument when you initialize `LogitGetter`, then during the forward pass, it decides whether or not to transpose the weight matrix based on the shape of the input embeddings and the shape of the weight matrix. If the weight matrix is square, then transposing and not transposing both work, so it can't make a correct decision.
> 
> If you want to avoid this ambiguity, you can explicitly set the transpose flag, but it depends on the loss function:
> 
> ```python
> # for ArcFaceLoss, CosFaceLoss, SphereFaceLoss, LargeMarginSoftmaxLoss, NormalizedSoftmaxLoss
> logit_getter = LogitGetter(classifier = loss_func, transpose = True)
> 
> # for ProxyAnchorLoss, ProxyNCALoss
> logit_getter = LogitGetter(classifier = loss_func, transpose = False)
> ```
> 
> Obviously this isn't great because you have to know whether or not to transpose the weight matrix. I've created an issue to address this: #256

Thank you - I understand now.

> 
> > * How are the logits exactly calculated? I can see that this should use the trainable parameters of the loss function, but I was unable to transverse through the code to locate where the logits are calculated for example for LargeMargin/ArcFace loss (I mean, it should be in a classifier layer I would have thought?). Any pointers to this would be great.
> 
> `LogitGetter` contains a copy of the weight matrix of the loss function. When you pass in embeddings, it does `distance(embeddings, weight_matrix)`, where `distance` is the distance metric used in the loss function. For example, `ArcFaceLoss` uses cosine similarity, and `NormalizedSoftmaxLoss` uses the dot product. If you want to use a different distance metric, you can pass it in as `distance` when you initialize `LogitGetter`.

Thank you.

I have one other question related to logitGetter. So, I would like to `jit.trace` my model, but, I would like the logits via logitGetter as the rsult of this `trace`. How does one go about this? Is this as simple as just including the logitGetter directly in the model? I havent tried it yet, but I thought I'd ask first if there are any `gotchas` in doing this before making an attempt.

**Comment by KevinMusgrave** [comment]
I've never used ```jit.trace``` so I don't have any thoughts on that unfortunately 😄
